### PR TITLE
Starting to sketch out IDE-level plugin modularity

### PR DESCRIPTION
**Pull request by alanz**
Address #25

Currently WIP

- [x] formatters
- [x] hover providers
- [x] code action providers
- [x] diagnostic providers
- [x] symbols
- [x] code lens provider
- [x] completions

### CODE REVIEW

**Comment by alanz** [comment]
It seems diagnostics work by generating a rule, and custom `RuleResult`, so get generated via the graph update process.

**Comment by ndmitchell** [comment]
I'm afraid I don't see how this works. I thought the PartialHandlers couldn't be combined? But you seem to be using them as is? Where does the merging for code lens happen, for example?

**Comment by alanz** [comment]
@ndmitchell this is in a very incomplete state at the moment, and I had to stop for a week. Working on it at the moment, will let you know when it makes sense to look at it again.

File: src/Ide/Plugin.hs [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        justs (p, Just x)  = [(p, x)]
+        justs (_, Nothing) = []
+
+        ls = Map.toList (ipMap mp)
+
+        mkPlugin :: ([(PluginId, b)] -> t) -> (PluginDescriptor -> Maybe b) -> t
+        mkPlugin maker selector
+            = maker $ concatMap (\(pid, p) -> justs (pid, selector p)) ls
+
+
```
Comment (alanz) [inline src/Ide/Plugin.hs]: This is the key part of this approach.

**Comment by alanz** [comment]
ping @bubba re https://github.com/haskell/haskell-language-server/pull/45/commits/e47361202119883f83df72f352323c8402e00bf0

**Comment by alanz** [comment]
This is by no means polished, but provides a skeleton for how we should proceed. My intention is to merge it soon, so it provides a context for others to be able to build.

File: haskell-language-server.cabal [hunk 2]
```diff
@@ -59,12 +65,14 @@ library
...
     , hie-bios >= 0.4
     , hslogger
+    , lens
     , optparse-simple
     , process
     , regex-tdfa >= 1.3.1.0
     , shake >= 0.17.5
     , text
     , transformers
+    , unix
```
Comment (jneira) [inline haskell-language-server.cabal]: ugh, this one make it unbuildable in windows 😄

Comment (alanz) [reply haskell-language-server.cabal]: what workarounds are there?

Comment (alanz) [reply haskell-language-server.cabal]: sorted, copied the approach from hie-plugin-api

**Comment by ndmitchell** [comment]
The basic idea looks reasonable. I think a `defaultDescriptor` would be a good idea so people don't have to list all the types of plugins they don't provide. Why does the plugin id get passed to every method in the descriptor? That seemed unnecessary as if you really needed it you could just put it in the closure. Other than that, seems pretty sensible. I didn't review all the changes in detail, more just tried to get a high-level sense of the design.

**Comment by alanz** [comment]
> Why does the plugin id get passed to every method in the descriptor?

We use the plugin id when constructing a command name to be sent to the client. These end up being the concatenation of the server process id, the plugin id, and the exposed command.

For me the critical thing is that the plugin id can be configured from the Main module when assembling a list of plugins for the particular instantiation, so we do not have to worry about command name clashes.  If there is some simpler way to do this, we can move in that direction.

This mirrors the approach we are currently taking in hie, so also makes it easier to port existing plugins.

**Pull request merged**